### PR TITLE
[JENKINS-14789] Rework periodic agent check

### PR DESCRIPTION
**Pull request by das7pad**
See [JENKINS-14789](https://issues.jenkins.io/browse/JENKINS-14789) and [this old thread from the mailing list](https://groups.google.com/g/jenkinsci-users/c/Hl4bE0mVYb4/m/OHbUFy8FEj0J).

This pull request is split into two parts: a set of bug fixes and a new
 system property for the check interval.

##### Bug fixes

Three bugs are addressed:
- Fluctuating loop invocations (caused by high CPU load) can lead to
  barely missed cycles within O(ms). Which in turn leads to slower
  startup times for agents (delay of up-to 2min instead of up-to 1min).
  Fix: Track the delay as the number of to-be-skipped loop ticks.
- High teardown delays can lead to extended periods of unavailable
  capacity when nodes are crashing during a teardown period.
  Fix: A changed online state also triggers a check now.
- `CloudSlaveRetentionStrategy` tracks the agent timeout in milliseconds
  internally and returned the nextCheck delay as a tenth of that.
  The nextCheck delay is consumed in minutes, leading to the max delay
  of one hour being used.
  Fix: Convert the delay from milliseconds to minutes.

##### New system property for the check interval

Administrators can trade elevated CPU usage from frequent agent checks
for responsiveness to capacity needs.

Per the docs on [`RetentionStrategy#check`](https://github.com/jenkinsci/jenkins/blob/c22cc07680cba8f8a401301a7712a779d4679796/core/src/main/java/hudson/slaves/RetentionStrategy.java#L61-L62), it is OK to rechecked earlier
or later than requested. Make the startup/teardown more responsive and
use the requested minutes as multiples of the configured interval.
The default interval is 60 seconds, which retains the original behavior.

Testing: Add a ssh-agent with an 'in demand delay' of `0` and specify
`-Dhudson.slaves.ComputerRetentionWork.checkIntervalSeconds=10` on the
command line. See a reasonably fast startup time of <10s.

In a follow-up we could move the system property into the GUI.

_I am a first-time contributor and could need some help with that._

----

I am not sure whether we can change the default check interval from one
 minute to a lower value like ten seconds.
This could lead to high CPU usage for setups with a long build queue `M`
 and many agents `N` as the runtime of the common in-demand strategies'
 check is `O(M * N^2)` and custom strategies might have even worse
 runtime characteristics.
A system property seems to be the safest approach to get this shipped.
 If we do decide to change the default interval, it could be an escape
 hatch for such big setups to bump it back up.

----

<details>
<summary> Manual testing performed </summary>

I gave this a good test in my (small) infra. It consists of a single
 controller and a dozen bash/ssh agents on different VMs. Each agent is
 configured with various labels, `inDemandDelay=0` (or `=1`) and
 `idleDelay=5`.
The newly added system property is set to `10` seconds.
A range of jobs sporting `agent any` and different combinations of
 labels was started. All workloads received their requested agents
 within <10s or 1min + <10s as configured. The idle agents were taken
 down after the configured 5min + <10s.
Manually started agents were taken down after 5min + <10s as expected.
Finally, I tested the restart of idle crashed agents with success. These
 steps are the reproducer for the 2nd fixed bug:
Start all the matching agents for some job A via the GUI, wait 2min for
 the idleDelay to kick in reliably (with either checkInterval=1min or
 10s), kill all the ssh-sessions, trigger a run for job A, see agents
 restarting in <10s.
The delay for agent restarts is about 3min before the bug fixes.
</details>

### Proposed changelog entries

* Rework the periodic agent check. Add a new system property `hudson.slaves.ComputerRetentionWork.checkIntervalSeconds` for (re)starting agents with a delay of less than one minute.

### Proposed upgrade guidelines

Please try out the new system property and report back whether you notice a significant increase in CPU usage.

Debian service default: `JAVA_ARGS=-Dhudson.slaves.ComputerRetentionWork.checkIntervalSeconds=10`
Docker: `docker run ... --env JAVA_OPTS=-Dhudson.slaves.ComputerRetentionWork.checkIntervalSeconds=10 jenkins/jenkins:jdk11`

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
    * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
  **Any pointers would be greatly appreciated. I could not find any 
   existing tests for the logic with a grep for `ComputerRetentionWork`/
   `computerRetention`/`doRun(`/ `run(`/`check(` in `test/`.**
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

N/A

### Maintainer checklist

Before the changes are marked as `ready-for-merge`:

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/5725"></a>

### CODE REVIEW

File: core/src/main/java/hudson/slaves/CloudSlaveRetentionStrategy.java [hunk 1]
```diff
@@ -38,7 +38,7 @@ public long check(T c) {
                 }
             }
         }
-        return checkCycle();
+        return TimeUnit.MILLISECONDS.toMinutes(checkCycle());
```
Comment (res0nance) [inline core/src/main/java/hudson/slaves/CloudSlaveRetentionStrategy.java:41]: According to the docs https://github.com/jenkinsci/jenkins/blob/ece17e043101dba0656a12266378e80328dd2592/core/src/main/java/hudson/slaves/RetentionStrategy.java#L61-L62

This change is correct

File: core/src/main/java/hudson/slaves/ComputerRetentionWork.java [hunk 2]
```diff
@@ -47,31 +48,53 @@
...
+     * Administrators can trade elevated CPU usage from frequent agent checks for responsiveness to capacity needs.
+     *
+     * Per the docs of {@link RetentionStrategy#check}, it is OK to rechecked earlier or later than requested.
+     * Make the startup/teardown more responsive and use the requested minutes as multiples of the configured interval.
+     * Setting the interval to 60 seconds (which is the default), retains the original behavior and
+     *  one has to wait up-to one minute for an in-demand agent (with inDemandDelay=0min) to start.
+     * Setting the interval to 10 seconds provides a more responsive behavior towards load changes and
+     *  one has to wait up-to ten seconds for an in-demand agent (with inDemandDelay=0min) to start.
+     */
+    private static final long checkIntervalSeconds = SystemProperties.getLong(ComputerRetentionWork.class.getName() + ".checkIntervalSeconds", TimeUnit.MINUTES.toSeconds(1));
```
Comment (res0nance) [inline core/src/main/java/hudson/slaves/ComputerRetentionWork.java:64]: This if poorly set can lead jenkins to perform poorly due to contention for the queue lock

**Comment by res0nance** [comment]
I'm also interested in what cloud you use, because I'm curious as to why your agents don't just pickup work as they get started but rely on retention strategy being called. 

A started agent should also call retention strategy start() immediately. IIRC.

**Comment by das7pad** [comment]
> I'm also interested in what cloud you use, because I'm curious as to why your agents don't just pickup work as they get started but rely on retention strategy being called.
> 
> A started agent should also call retention strategy start() immediately. IIRC.

I am using a static set of VMs for the agents. The VMs are performing other tasks for most of the day and pick up CI work as needed. I do not see a need for keeping a persistent connection from the master node to all of them running when no work is available. Using different `inDemandDelay`s also allows me to prefer faster VMs over slower ones :slightly_smiling_face: .

File: core/src/main/java/hudson/slaves/ComputerRetentionWork.java (hunk 2 above)
Comment (das7pad) [reply core/src/main/java/hudson/slaves/ComputerRetentionWork.java:64]: Agree. WDYT about putting a lower limit of three seconds in place?

**Comment by das7pad** [comment]
I've rebased this PR onto the last weekly release to fix conflicts with #5732 (Sort imports).

_(A cursory test of the happy paths checks out: startup/teardown delay remains <10s above the specified inDemand/idleDelay.)_

**Comment by jimklimov** [comment]
Regarding integration with GUI, you can look at my recent PR #5764 as an example that added a field to individual build agent configuration, in the onDemand area. I suspect your PR could benefit from having a per-agent tunable nearby. And the goals of my PR (sequentially using different agents on same host) would benefit from faster hand-over thanks to your PR :)

**Review by basil — COMMENTED** [review]
@das7pad This seems like two separate changes, neither of which are complete.

Can you start by splitting the `checkIntervalSeconds` into a separate PR, along with a GUI to be able to customize the option? After that is merged, we can discuss the remainder of this PR.

**Comment by basil** [comment]
Hi @das7pad, do you have any interest in continuing with this PR?

**Comment by das7pad** [comment]
> Hi @das7pad, do you have any interest in continuing with this PR?

Yes. I've opened https://github.com/jenkinsci/jenkins/pull/7037 with the config option in the GUI.

**Comment by das7pad** [comment]
Superseded by https://github.com/jenkinsci/jenkins/pull/7037

**Pull request closed without merging**